PR: https://github.com/odoo/odoo/pull/235438

From: 19ab8c7344c85e123ac1f8b5dba87769abc3a8fd
From: romc-odoo
Date: 2025-11-28 09:30:08

Breaking data model changes scores: del:2 + add:0, change matches:
-    hr_presence_state_display = fields.Selection([
-    hr_presence_state_display = fields.Selection([

Total Changes: 28

[IMP] hr,hr_presence: allow filtering and grouping by presence state

The stored presence state field is removed as it did not always contain the correct state value.

This commit introduces the possibility to filter and group with the computed hr presence state.
As this is a computed non-stored field, the implementation has poor performances.
Even if the search method of the field is implemented, it should not be used outside of search views domains.

task-5082744

closes odoo/odoo#235438

Related: odoo/upgrade#8940
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_presence/models/hr_employee.py
+++ b/addons/hr_presence/models/hr_employee.py
@@ -18,14 +18,6 @@ class HrEmployee(models.Model):
     manually_set_present = fields.Boolean(default=False)
     manually_set_presence = fields.Boolean(default=False)
 
-    # Stored field used in the presence kanban reporting view
-    # to allow group by state.
-    hr_presence_state_display = fields.Selection([
-        ('out_of_working_hour', 'Off-Hours'),
-        ('present', 'Present'),
-        ('absent', 'Absent'),
-        ], default='out_of_working_hour')
-
     @api.model
     def _check_presence(self):
         company = self.env.company
@@ -73,9 +65,6 @@ class HrEmployee(models.Model):
 
         company.sudo().hr_presence_last_compute_date = Datetime.now()
 
-        for employee in all_employees:
-            employee.hr_presence_state_display = employee.hr_presence_state
-
     def get_presence_server_action_data(self):
         server_action_xmlids = [
             'action_hr_employee_presence_present',
@@ -94,7 +83,6 @@ class HrEmployee(models.Model):
         self.write({
             'manually_set_present': state,
             'manually_set_presence': True,
-            "hr_presence_state_display": 'present' if state else 'absent',
         })
 
     def action_set_present(self):
@@ -103,11 +91,6 @@ class HrEmployee(models.Model):
     def action_set_absent(self):
         self._action_set_manual_presence(False)
 
-    def write(self, vals):
-        if vals.get('hr_presence_state_display') == 'present':
-            vals['manually_set_present'] = True
-        return super().write(vals)
-
     def action_open_leave_request(self):
         if len(self) == 1:
             model = 'hr.leave'
@@ -158,16 +141,12 @@ Thank you for your prompt attention to this matter.""")
             "target": "new",
         }
 
-    @api.depends("user_id.im_status", "hr_presence_state_display")
+    @api.depends("user_id.im_status")
     def _compute_presence_state(self):
         super()._compute_presence_state()
         company = self.env.company
         working_now_list = self._get_employee_working_now()
         for employee in self:
-            if employee.manually_set_presence:
-                employee.hr_presence_state = employee.hr_presence_state_display
-                continue
-
             if not employee.company_id.hr_presence_control_email and not employee.company_id.hr_presence_control_ip:
                 continue
             if company.hr_presence_last_compute_date and employee.id in working_now_list and \

--- a/addons/hr_presence/models/hr_employee_public.py
+++ b/addons/hr_presence/models/hr_employee_public.py
@@ -10,8 +10,3 @@ class HrEmployeePublic(models.Model):
     ip_connected = fields.Boolean(default=False)
     manually_set_present = fields.Boolean(default=False)
     manually_set_presence = fields.Boolean(default=False)
-    hr_presence_state_display = fields.Selection([
-        ('out_of_working_hour', 'Off-Hours'),
-        ('present', 'Present'),
-        ('absent', 'Absent'),
-        ], default='out_of_working_hour')
